microprofile/metrics/src/main/java/io/helidon/microprofile/metrics/InterceptInfo.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package io.helidon.microprofile.metrics;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Executable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * Records information about an intercepted method.
+ * <p>
+ *     Specifically:
+ * <ul>
+ *     <li>the work items to be updated when the corresponding method is intercepted, organized by the annotation class that
+ *     gave rise to each work item; and</li>
+ *     <li>the {@code InterceptRunner} to use in updating the work items and invoking the method or constructor.</li>
+ * </ul>
+ * </p>
+ *
+ * @param <T> base type of the work items handled by the interceptor represented by this instance
+ */
+class InterceptInfo<T> {
+
+    private final InterceptRunner runner;
+
+    private final Map<Class<?>, Collection<T>> workItemsByAnnotationType = new HashMap<>();
+
+    /**
+     * Creates a new instance based on the provided {@code Executable}.
+     *
+     * @param executable the constructor or method subject to interception
+     * @return the new instance
+     */
+    static <T> InterceptInfo<T> create(Executable executable) {
+        return new InterceptInfo<>(InterceptRunnerImpl.create(executable));
+    }
+
+    private InterceptInfo(InterceptRunner runner) {
+        this.runner = runner;
+    }
+
+    InterceptRunner runner() {
+        return runner;
+    }
+
+    Supplier<Iterable<T>> workItems(Class<? extends Annotation> annotationType) {
+        /*
+         * Build a supplier of the iterable, because before-and-after runners will need to process the work items twice so we need
+         *  to give the runner a supplier.
+         */
+        return () -> workItemsByAnnotationType.get(annotationType);
+    }
+
+    /**
+     * Adds a work item to this info, identifying the interceptor type that will update this work item.
+     * @param annotationClass type of the interceptor
+     * @param workItem the newly-created workItem
+     */
+    void addWorkItem(Class<? extends Annotation> annotationClass, T workItem) {
```

Comment:
I missed `annotationClass` in a naming consistency check. It should be `annotationType`. 

Yes, `workItem` is generic, intentionally so. 

In this specific use in this PR, each work item (`MetricWorkItem` added in this PR) consists of a `MetricID` (for checking to make sure the metric has not been removed from the registry) and a reference to the `Metric` to be updated around an `Executable` invocation.

But In a parallel PR adding Micrometer support, this class migrates to the new `service-common/rest-cdi` module because it'll be used for both MP metrics and Micrometer meters. Plus, any other Helidon module that uses interceptors could reuse this and the other common behavior that's moving to the new module. And there's no telling what "work" other modules' interceptors might need to do when they run. 

"Work" seemed like a generic (which is not necessarily a bad thing, considering) but expressive term to adopt for this, but again I'm open to suggestions!